src/usdb_syncer/usdb_scraper.py

```
@@ -40,95 +44,235 @@
     r'<td onclick="show_detail\(\d+\)">(?P<rating>.*?)</td>\n'
     r'<td onclick="show_detail\(\d+\)">(?P<views>.*?)</td>'
 )
-WELCOME_REGEX = re.compile(
-    r"<td class='row3' colspan='2'>\s*<span class='gen'>([^<]+) <b>([^<]+)</b>"
-)
 
 
-def establish_usdb_login(session: Session) -> bool:
-    """Tries to log in to USDB if necessary. Returns final login status."""
-    if user := get_logged_in_usdb_user(session):
-        logger.info(f"Using existing login of USDB user '{user}'.")
-        return True
-    if (auth := settings.get_usdb_auth())[0] and auth[1]:
-        if login_to_usdb(session, *auth):
-            logger.info(f"Successfully logged in to USDB with user '{auth[0]}'.")
+class UsdbSession(SyncerSession[str]):
+    """Session class for USDB."""
+
+    def __init__(self, **kwargs: dict) -> None:
+        super().__init__(base_url=Usdb.BASE_URL, **kwargs)
+        self._username: str = ""
+        self._connected: bool = False
+
+    @override
+    def handle_response(self, response: Response) -> str:
+        """Handle response."""
+        if not self.connected:
+            self.connected = True
+        if not response.ok:
+            logger.debug(
+                f"Request failed with status code {response.status_code} for URL: "
+                f"{response.url}"
+            )
+            return ""
+        return normalize(response.text)
+
+    @override
+    def conn_failed(self, url: str, error: RequestException) -> str:
+        """Handle connection failure."""
+        self.connected = False
```

Swallowing all these exceptions that were raised before will have some unforeseen consequences (below is just the one I noticed immediately).

---

True, I forgot to handle this. I suppose there are two solutions:

1. Raise the exception here
2. Add an `if text := self.get` for every get request and deal with the error explicitly along the call stack. Probably involves changing `song_loader` pretty extensively.

Any other ideas? I'm not too hung up on this method in general, but I'm also not a fan of the previous solution of just erroring. The pain of web requests in general, I guess.

---

The syncer is built to deal with most unexpected exception. To handle things more gracefully, I don't see an alternative to catching exceptions on a case-by-case basis.
On a related note, I don't think the session class is the ideal level to define response handling and type. They could easily differ between requests.